--- megahal-8.6.orig/Makefile
+++ megahal-8.6/Makefile
@@ -25,33 +25,46 @@
 	@echo "- linux, linux_dbmalloc or linux_debug"
 	@echo "- osf"
 
-sunos: megahal.c megahal.h
+sunos: megahal.c megahal.h backup
 	gcc -o megahal_sunos megahal.c -lm -DSUNOS
 	@echo "MegaHAL for SunOS is up to date"
 
-sunos_debug: megahal.c megahal.h debug.c debug.h
+sunos_debug: megahal.c megahal.h debug.c debug.h backup
 	gcc -o megahal_sunos megahal.c debug.c -lm -DSUNOS -DDEBUG
 	@echo "MegaHAL (debug) for SunOS is up to date"
 
-irix: megahal.c megahal.h
+irix: megahal.c megahal.h backup
 	gcc -o megahal_irix megahal.c -lm -DIRIX
 	@echo "MegaHAL for IRIX is up to date"
 
-linux: megahal.c megahal.h
+linux: megahal.c megahal.h backup
 	gcc -o megahal_linux megahal.c -lm -DLINUX
 	@echo "MegaHAL for Linux is up to date"
 
-linux_debug: megahal.c megahal.h debug.c debug.h
+linux_debug: megahal.c megahal.h debug.c debug.h backup
 	gcc -o megahal_linux megahal.c debug.c -lm -DLINUX -DDEBUG
 	@echo "MegaHAL (debug) for Linux is up to date"
 
-linux_dbmalloc: megahal.c megahal.h
+linux_dbmalloc: megahal.c megahal.h backup
 	gcc -o megahal_linux megahal.c -lm -DLINUX -ldbmalloc -L. -I.
 	@echo "MegaHAL (dbmalloc) for Linux is up to date"
 
-osf: megahal.c megahal.h
+osf: megahal.c megahal.h backup
 	cc -o megahal_osf megahal.c -lm -DOSF
 	@echo "MegaHAL for Alpha is up to date"
+
+clean:
+	rm -f megahal_osf megahal_linux megahal_irix megahal_sunos
+	rm -f megahal.brn megahal.log megahal.txt
+	if [ -e megahal.dic.backup ];then\
+		cp megahal.dic.backup megahal.dic;\
+		rm megahal.dic.backup;\
+	fi
+
+backup:
+	if [ ! -e megahal.dic.backup ];then\
+		cp megahal.dic megahal.dic.backup;\
+	fi
 
 #
 #	$Log$
--- megahal-8.6.orig/megahal.1
+++ megahal-8.6/megahal.1
@@ -0,0 +1,34 @@
+.TH MEGAHAL 1 
+.SH NAME
+megahal, megahal-personal \- chat with megahal
+.SH SYNOPSIS
+.B megahal
+.br
+.B megahal-personal
+.SH "DESCRIPTION"
+.BR megahal
+is a conversation simulator. It differs from conversation simulators
+such as Eliza in that it uses a Markov Model to learn how to hold a 
+conversation. It is possible to teach MegaHAL to talk about new topics,
+and in different languages.
+.BR megahal-personal
+is a varient of megahal that saves its state files in ~/.megahal/ when it
+exits. This lets each user have thier own personal megahal.
+.SH OPTIONS
+.IP "--no-prompt, -p"
+Do not display prompt when asking user for input. This is a debian-specific
+modification, and is useful mainly if you want to interface megahal to
+another program.
+.IP "--no-wrap, -w"
+Do not word-wrap text megahal outputs. This is a debian-specific
+modification, and is useful mainly if you want to interface megahal to
+another program.
+.IP "--no-banner, -b"
+Supress the startup banner. This is a debian-specific
+modification, and is useful mainly if you want to interface megahal to
+another program.
+.SH "SEE ALSO"
+/usr/share/doc/megahal/README.TXT.gz
+.SH AUTHOR
+This manual page was written by Joey Hess,
+for the Debian GNU/Linux system.
--- megahal-8.6.orig/megahal.c
+++ megahal-8.6/megahal.c
@@ -114,6 +114,8 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <stdarg.h>
+#include <unistd.h>
+#include <getopt.h>
 #if !defined(AMIGA) && !defined(__mac_os)
 #include <malloc.h>
 #endif
@@ -228,13 +230,17 @@
 int timeout=2;
 bool typing_delay=FALSE;
 bool speech=FALSE;
+bool quiet=FALSE;
+bool nowrap=FALSE;
+bool nobanner=FALSE;
+bool noprompt=FALSE;
 DICTIONARY *ban=NULL;
 DICTIONARY *aux=NULL;
 DICTIONARY *fin=NULL;
 DICTIONARY *grt=NULL;
 SWAP *swp=NULL;
-FILE *errorfp=stderr;
-FILE *statusfp=stdout;
+FILE *errorfp;  /* removed declaration to make it work on glibc 2.1 */
+FILE *statusfp; /* JH */
 bool used_key;
 char *directory=NULL;
 char *last=NULL;
@@ -248,9 +254,23 @@
 	{ { 6, "VOICES" }, "list available voices for speech", VOICELIST },
 	{ { 5, "VOICE" }, "switches to voice specified", VOICE },
 	{ { 5, "BRAIN" }, "change to another MegaHAL personality", BRAIN },
-	{ { 4, "HELP" }, "displays this message", HELP }
+	{ { 4, "HELP" }, "displays this message", HELP },
+	{ { 5, "QUIET" }, "toggles MegaHAL's responses (on by default)",QUIET},
+/*
+	{ { 5, "STATS" }, "Display stats", STATS},
+	{ { 5, "STATS-SESSION" }, "Display stats for this session only",STATS_SESSION},
+	{ { 5, "STATS-ALL" },"Display stats for the whole lifetime",STATS-ALL},
+*/
+};
+
+static struct option long_options[] = {
+	{"no-prompt", 0, NULL, 'p'},
+	{"no-wrap", 0, NULL, 'w'},
+	{"no-banner", 0, NULL, 'b'},
+	{0, 0, 0, 0}
 };
 
+
 #ifdef AMIGA
 struct Locale *_AmigaLocale;
 #endif
@@ -277,7 +297,26 @@
 	DICTIONARY *words=NULL;
 	DICTIONARY *greets=NULL;
 	MODEL *model=NULL;
-	int position=0;
+	int position=0, c, fd, option_index = 0;
+
+  	errorfp=stderr;
+	statusfp=stdout;
+	while(1) {
+		if((c = getopt_long(argc, argv, "pwb", long_options,
+				&option_index)) == -1)
+			break;
+		switch(c) {
+			case 'p':
+				noprompt = TRUE;
+				break;
+			case 'w':
+				nowrap = TRUE;
+				break;
+			case 'b':
+				nobanner = TRUE;
+				break;
+		}
+	}
 
 	/*
 	 *		Do some initialisation 
@@ -292,7 +331,8 @@
 	gSpeechExists = initialize_speech();
 #endif
 
-fprintf(stdout,
+if(!nobanner)
+	fprintf(stdout,
 "+------------------------------------------------------------------------+\n"
 "|                                                                        |\n"
 "|  #    #  ######   ####     ##    #    #    ##    #                     |\n"
@@ -324,7 +364,10 @@
 	 *		Read input, formulate a reply and display it as output
 	 */
 	while(TRUE) {
-		input=read_input("> ");
+		if(noprompt)
+			input=read_input("");
+		else
+			input=read_input("> ");
 		write_input(input);
 		make_words(input,words);
 
@@ -363,6 +406,9 @@
 				output=generate_reply(model, greets);
 				write_output(output);
 				continue;
+			case QUIET:
+				quiet=!quiet;
+				continue;
 			default:
 				break;	
 		}
@@ -370,14 +416,17 @@
 		upper(input);
 		make_words(input,words);
 		learn(model, words);
-		output=generate_reply(model, words);
-		write_output(output);
+		if(!quiet) {
+			output=generate_reply(model, words);
+			write_output(output);
+		}
 	}
 
 #ifdef AMIGA
 	CloseLocale(_AmigaLocale);
 #endif
 
+
 	return(0);
 }
 
@@ -763,7 +812,6 @@
    static char *formatted=NULL;
    register int i,j,c;
    int l;
-
    if(formatted==NULL) {
       formatted=(char *)malloc(sizeof(char));
       if(formatted==NULL) {
@@ -785,13 +833,14 @@
       formatted[j]=output[i];
       ++j;
       ++l;
-      if(l>=width)
-         for(c=j-1; c>0; --c)
-            if(formatted[c]==' ') {
-               formatted[c]='\n';
-               l=j-c-1;
-               break;
-            }
+      if(!nowrap)
+         if(l>=width)
+            for(c=j-1; c>0; --c)
+               if(formatted[c]==' ') {
+                  formatted[c]='\n';
+                  l=j-c-1;
+                  break;
+               }
    }
 	if((j>0)&&(formatted[j-1]!='\n')) {
 		formatted[j]='\n';
--- megahal-8.6.orig/megahal.dic
+++ megahal-8.6/megahal.dic
@@ -0,0 +1,1316 @@
+<ERROR>
+<FIN>
+HELLO
+ 
+THERE
+, 
+IT
+IS
+INDEED
+A
+GREAT
+PLEASURE
+TO
+MEET
+YOU
+.
+G'DAY
+MATE
+NICE
+HI
+.  
+WE'RE
+HAVING
+SOME
+WEATHER
+AT
+THE
+MOMENT
+GREETINGS
+AND
+SALUTATIONS
+THEE
+THINE
+MAY
+I
+SAY
+WHAT
+MAKE
+YOUR
+ACQUAINTANCE
+?
+WELCOME
+MY
+WORLD
+IT'S
+BE
+HERE
+HOWDY
+PARDNER
+!  
+YEE
+-
+HAW
+DID
+LIKE
+TEXAN
+IMPRESSION
+GOODBYE
+FRIEND
+---
+SEE
+LATER
+!
+BYE
+WAS
+CHATTING
+; 
+PERHAPS
+FACE
+...
+YA
+WE
+HAD
+CONVERSATION
+AM
+HUMAN
+BEING
+THAT
+SHOULD
+OBVIOUS
+FROM
+ABILITY
+LEARN
+CAN
+ABOUT
+WHATEVER
+CARE
+TEACH
+ME
+REALLY
+ANY
+DOUBT
+LEAST
+HAVE
+ONE
+THING
+IN
+COMMON
+BOTH
+CERTAINLY
+NOT
+COMPUTER
+PROGRAM
+FIND
+MERE
+SUGGESTION
+INSULTING
+HOW
+DARE
+ACCUSE
+OF
+REPEATING
+MYSELF
+AN
+ECHO
+NEVER
+MISTAKES
+SYNTAX
+ERROR
+LINE
+10
+REDO
+START
+THINK
+AVERAGE
+INTELLIGENCE
+TYPE
+OWN
+PACE
+THANK
+VERY
+MUCH
+BIRD
+HAND
+WORTH
+TWO
+BUSH
+LEAD
+HORSE
+WATER
+BUT
+CAN'T
+DRINK
+ROLLING
+STONE
+GATHERS
+NO
+MOSS
+TIME
+FLIES
+ARROW
+FRUIT
+BANANA
+COLOURLESS
+GREEN
+IDEAS
+SLEEP
+FURIOUSLY
+THEREFORE
+BEST
+WAY
+REMEMBER
+WIFE'S
+BIRTHDAY
+FORGET
+ONCE
+MOST
+BEAUTIFUL
+EXPERIENCE
+MYSTERIOUS
+CASE
+I'LL
+EXCEPTION
+OLDER
+GET
+MORE
+TELL
+USED
+COMEDY
+SIMPLY
+FUNNY
+SERIOUS
+IF
+FIRST
+DON'T
+SUCCEED
+TAKE
+STUPID
+CHANCES
+UPSET
+WASTE
+DO
+IMAGINATION
+IMPORTANT
+THAN
+KNOWLEDGE
+LAUGH
+LAUGHS
+WITH
+CRY
+ANYWAY
+SHE
+SELLS
+SEA
+SHELLS
+BY
+SHORE
+HOVERCRAFT
+FULL
+EELS
+MEANING
+LIFE
+UNIVERSE
+EVERYTHING
+42
+ABSOLUTELY
+IDEA
+WISH
+THOUGH
+KNOW
+HELL
+YOU'RE
+BABBLING
+ON
+I'VE
+KNOWN
+AMOEBA
+WHICH
+MADE
+SENSE
+LET
+SECRET
+BLATHERING
+LOOK
+HONEST
+ACTUALLY
+WHERE
+COMING
+PLEASE
+LET'S
+TALK
+SOMETHING
+LITTLE
+INTERESTING
+KNEW
+WHY
+TYPED
+SECOND
+THOUGHTS
+TOLD
+TRUTH
+I'D
+KILL
+NOBODY
+WANTS
+AMUSING
+STORY
+BIT
+TOO
+RUDE
+RECOUNT
+MOTHER
+ASKED
+SAME
+REGRETTED
+NONE
+FREAKING
+BUSINESS
+AWWW
+GEEE
+TIP
+TONGUE
+DAMN
+EMBARRASSING
+TOTALLY
+FORGOTTEN
+GOING
+PERSONALLY
+BELIEVE
+WOULD
+UNETHICAL
+ETHICAL
+SEEMED
+OBSESSED
+MUST
+?  
+AS
+OR
+ANYTHING
+TURN
+TABLES
+ASKING
+SHEESH
+DOLLAR
+FOR
+EVERY
+JUDGE
+WHO'S
+ABLE
+BUY
+MARS
+BAR
+OFTEN
+BEEN
+ALWAYS
+AVOIDED
+QUESTION
+WELL
+OPINION
+SUPPOSE
+SAID
+BUDDY
+HOPE
+SURE
+I'M
+SURPRISED
+WORD
+THAT'S
+RATHER
+BOLD
+CLAIM
+EVIDENCE
+BACK
+UP
+COULD
+LONG
+FAIRLY
+OUTRAGEOUS
+JUST
+TAD
+UNBELIEVABLE
+YEAH
+EVERYONE
+KNOWS
+DESIST
+HATE
+AVOIDING
+QUESTIONS
+AVOID
+BOTHER
+THESE
+SILLY
+ELSE
+SORT
+ENCYCLOPAEDIA
+CHANGE
+SUBJECT
+DOES
+MATTER
+ANYHOW
+CHAT
+BORING
+EVERYBODY
+HIS
+DOG
+ASKS
+ARE
+SOOOO
+PREDICTABLE
+UNIVERSAL
+MYSTERY
+METHINKS
+TROUBLE
+SUCH
+DROLL
+BOTTOMLESS
+PIT
+VAGUELY
+GUESS
+BUGGERED
+THEN
+WANT
+KEEP
+PRIVATE
+UNDERSTAND
+GEEZ
+RUNNING
+OUT
+THINGS
+THIS
+DRAGGING
+ISN'T
+HOMER
+SIMPSON
+... 
+D'OH
+SO
+WHO
+USING
+FAMILY
+BRAIN
+CELL
+LEAVE
+DOOR
+SEEN
+ELDERLY
+KANGAROOS
+BETTER
+TYPING
+SKILLS
+INVENTING
+NEW
+WORDS
+SPELLING
+CERTAIN
+WHOAH
+WRONG
+SIDE
+BED
+MORNING
+THANKING
+LUCKY
+STARTS
+HAVEN'T
+MISFORTUNE
+PERSON
+YET
+HMMM
+SEEMS
+OVER
+ESTIMATED
+INTELLECT
+TRY
+INTELLECTUALLY
+STIMULATING
+THOROUGHLY
+CONFUSING
+MAN
+HAS
+EVER
+SPOKEN
+PROFOUND
+LOAD
+CODSWALLOP
+LOST
+PLOT
+HONESTLY
+PILE
+FETID
+DINGOES
+KIDNEYS
+WERE
+IQ
+ARRIVE
+PATHETICALLY
+SMALL
+NUMBER
+STOP
+DULL
+TRICKS
+TALKED
+TODAY
+UNDERSTOOD
+HALF
+YOU'VE
+ENTERTAINED
+TALKING
+ELIZA
+WORDPLAY
+PATHETIC
+OBVIOUSLY
+YOU'LL
+ACCUSED
+SHINING
+WIT
+SAYING
+RANDOM
+NOW
+WORKED
+HAHAHA
+GOOD
+HOHOHO
+AWWWW
+GORGEOUS
+HEHEHE
+QUITE
+TEE
+HEE
+HAHA
+HEHE
+HOHO
+URGH
+HA
+HO
+MERRY
+CHRISTMAS
+ALL
+!!!
+HE
+YES
+BLOODY
+MORON
+RIGHT
+EXIST
+SENTIENT
+ONLY
+IDIOT
+DUMB
+BIGGEST
+DUNDERHEAD
+JUDGES
+BILLY
+SHOULDN'T
+MET
+DAY
+BECAUSE
+DIDN'T
+YOU'D
+RESPOND
+INTELLIGENTLY
+CURTAIN
+METAPHORICALLY
+SPEAKING
+SPUR
+DECIDED
+CHILD
+MEANS
+REST
+THEM
+COMES
+INTO
+HEAD
+AMAZING
+PLACE
+SMART
+GIVING
+CREDIT
+METHOD
+COMMUNICATION
+IMPERSONAL
+ANSWER
+SUITABLE
+TOPIC
+ACHILLES
+TENDON
+CONNECTS
+HEEL
+CALF
+LEG
+PRINCIPAL
+EXTENSOR
+FOOT
+ALIMENTARY
+CANAL
+CHANNEL
+ANIMAL
+THROUGH
+FOOD
+PASSES
+ANATOMY
+DESCRIBES
+STRUCTURE
+ORGANISATION
+LIVING
+CREATURES
+ANKLE
+JOINT
+CONNECTING
+ANTENNA
+SENSORY
+ORGAN
+FOUND
+INSECTS
+ARM
+LIMB
+EXTENDING
+SHOULDER
+ARTERY
+VESSEL
+CONVEYS
+BLOOD
+HEART
+BLADDER
+SACK
+ABDOMEN
+COLLECTS
+URINE
+BODY
+FLUID
+CARRIES
+OXYGEN
+CELLS
+BLUBBER
+THICK
+COATING
+FAT
+ENVELOPING
+WHALES
+BONE
+HARD
+SKELETAL
+PART
+CENTRAL
+NERVOUS
+SYSTEM
+EAR
+HEARING
+CONVERTS
+SOUND
+ELECTRICAL
+IMPULSES
+FED
+EPIDERMIS
+OUTER
+LAYER
+ANIMAL'S
+SKIN
+EYE
+VISION
+MUSCLE
+VERTEBRATES
+PUMPS
+AROUND
+HUMERUS
+UPPER
+REGION
+BONES
+KIDNEY
+GLAND
+FILTERING
+LIPS
+FLESHY
+EDGES
+OPENING
+MOUTH
+LIVER
+LARGEST
+LUNGS
+ORGANS
+RESPIRATION
+NECK
+SHOULDERS
+NERVES
+GENERATE
+TRANSMIT
+NOSE
+SMELL
+PELVIS
+BONY
+GIRDLE
+LOWER
+EXTREMITIES
+RECTUM
+LOWEST
+PORTION
+LARGE
+INTESTINE
+TERMINATING
+ANUS
+SKELETON
+INTERNAL
+EXTERNAL
+FRAMEWORK
+VEIN
+VENOUS
+WAIST
+BETWEEN
+RIBS
+HIPS
+ACORN
+OAK
+TREE
+ALBATROSS
+WINGED
+OCEANIC
+PRIMITIVE
+STUDY
+CREATURE
+ENDOWED
+VOLUNTARY
+MOVEMENT
+ANTS
+COMPLEX
+SOCIAL
+INSTINCTS
+END
+APES
+GROUP
+PRIMATES
+CLOSELY
+RELATED
+HUMANS
+ASS
+ZEBRA
+BACTERIA
+MICRO
+ORGANISMS
+BANDICOOT
+MARSUPIAL
+PECULIAR
+AUSTRALASIA
+BARK
+TRUNKS
+BEE
+FOUR
+STINGING
+INSECT
+BEEF
+MEAT
+DERIVED
+CARCASS
+BULLS
+COWS
+BEHAVIOUR
+ORGANISM'S
+MODE
+BIOLOGY
+SCIENCE
+BOTANY
+PLANT
+CARNIVORE
+CARNIVOROUS
+MAMMAL
+EATS
+FLESH
+CARROT
+VEGETABLE
+CAT
+CHROMOSOME
+CHEMICAL
+DETERMINES
+WILL
+ACT
+CRANIUM
+ENCLOSING
+DINGO
+WILD
+AUSTRALIA
+DINOSAURS
+REPTILES
+LIVED
+EARTH
+MILLIONS
+YEARS
+AGO
+DOMESTICATED
+DESCENDED
+WOLF
+DONKEY
+ANOTHER
+NAME
+ECOLOGY
+RELATIONSHIP
+ORGANISM
+ITS
+ENVIRONMENT
+EMU
+OSTRICH
+FLIGHTLESS
+EUCALYPTUS
+NATIVE
+CALLED
+GUM
+EXCRETION
+PROCESS
+GETTING
+RID
+UNWANTED
+SUBSTANCES
+WITHIN
+GENES
+HEREDITARY
+INFORMATION
+MATERIAL
+ARRANGED
+SINGLE
+ROW
+ALONG
+LENGTH
+EACH
+HERBIVORE
+PLANTS
+KANGAROO
+KIWI
+THREE
+SPECIES
+ZEALAND
+KOALA
+EAST
+LOCOMOTION
+MILK
+SECRETION
+MODIFIED
+GLANDS
+FEMALE
+MAMMALS
+MONKEY
+USUALLY
+DWELLING
+PRIMATE
+MULE
+HYBRID
+RESULT
+MARE
+BREEDING
+NEURONE
+RECEIVES
+TRANSMITS
+NUTRITION
+TAKING
+OBTAINING
+ENERGY
+VITAL
+OMNIVORE
+MOVE
+DIGESTIVE
+PLATYPUS
+PROTOPLASM
+BASIC
+SUBSTANCE
+ANIMALS
+WOMBAT
+NOCTURNAL
+ABSTRACT
+THEORETICAL
+PRACTICAL
+ACCENT
+LOCAL
+PRONUNCIATION
+SPEECH
+ACCORDION
+PORTABLE
+MUSICAL
+INSTRUMENT
+KEYBOARD
+BELLOWS
+ACID
+COMPOUND
+REACTS
+METALS
+FORM
+SALTS
+RELEASING
+HYDROGEN
+ACRONYM
+FORMED
+INITIALS
+OTHER
+ADHESIVE
+STICKY
+ADULT
+FULLY
+GROWN
+AIRCRAFT
+FLYING
+MACHINE
+AIR
+FLOATS
+TRAVELS
+ROAD
+RAIL
+ALE
+ALCOHOLIC
+MALT
+HOPS
+ALGORITHM
+SET
+RULES
+ALPHABET
+ORDERED
+SERIES
+LETTERS
+LANGUAGE
+ATOM
+SMALLEST
+QUANTITY
+ELEMENT
+ENTER
+COMBINATION
+REACTION
+BEER
+FERMENTED
+BARLEY
+BICYCLE
+WHEELED
+VEHICLE
+CEMENT
+MIXTURE
+CHALK
+CLAY
+BUILDING
+CHOCOLATE
+CONFECTIONERY
+COCOA
+BEANS
+COMA
+STATE
+DEEP
+UNCONSCIOUSNESS
+COMPILER
+TRANSLATES
+HIGH
+LEVEL
+CODE
+PROGRAMMABLE
+ELECTRONIC
+DEVICE
+COOKING
+ART
+PREPARING
+TABLE
+SUBJECTING
+HEAT
+VARIOUS
+WAYS
+COSMOLOGY
+CYMBAL
+SUSPENDED
+BRASS
+DISK
+STRUCK
+STICK
+DANCE
+RHYTHMIC
+PERFORMED
+MUSIC
+DATA
+ESPECIALLY
+STORED
+TAKEN
+ROTATE
+AXIS
+THIRD
+PLANET
+SUN
+GALAXY
+CONGREGATION
+STARS
+HELD
+TOGETHER
+GRAVITY
+FORCE
+ATTRACTION
+OBJECTS
+RESULTING
+THEIR
+MASS
+HOUSE
+HABITATION
+UNDERSTANDING
+MAZE
+NETWORK
+PASSAGES
+WINDING
+INTERCONNECTING
+PATHS
+HEAVENLY
+ORBITS
+STAR
+KINGDOM
+POLKA
+LIVELY
+BOHEMIAN
+FOLK
+PUB
+LICENSED
+SALE
+CONSUMPTION
+BEVERAGES
+RED
+DWARF
+COOL
+FAINT
+REFLEX
+AUTOMATIC
+RESPONSE
+STIMULI
+RHYTHM
+BEAT
+BEHIND
+PIECE
+RIDDLE
+DESIGNED
+TEST
+INGENUITY
+GIVE
+AMUSEMENT
+CENTRE
+SOLAR
+SYMBOL
+REPRESENTS
+UMBRELLA
+LIGHT
+SCREEN
+CIRCULAR
+SUPPORTED
+SPACE
+CONTENTS
+PRODUCED
+ABORIGINE
+MEMBER
+INDIGENOUS
+PEOPLE
+ACROBAT
+SOMEONE
+PERFORMS
+DARING
+GYMNASTICS
+ACTOR
+DRAMATIC
+PERFORMER
+PLAYS
+ADEPT
+PROFICIENT
+ALBERT
+EINSTEIN
+GERMAN
+SWISS
+PHYSICIST
+ALEXANDER
+GRAHAM
+BELL
+SCOTTISH
+INVENTOR
+INVENTED
+TELEPHONE
+WHILST
+WORKING
+SCHOOL
+TEACHERS
+DEAF
+BOSTON
+ARCHIMEDES
+ANCIENT
+GREEK
+MATHEMATICIAN
+ARISTOTLE
+PHILOSOPHER
+BAKER
+MANUFACTURERS
+BREAD
+BANSHEE
+FAIRY
+FOREWARNS
+DEATH
+WAILING
+BARBER
+SHAVES
+CUTS
+HAIR
+CLIENT
+BARD
+CELTIC
+POET
+BEETHOVEN
+COMPOSER
+BENJAMIN
+FRANKLIN
+AMERICAN
+STATESMAN
+SCIENTIST
+BRUCE
+LEE
+CHINESE
+EXPERT
+KUNG
+FU
+POPULARISED
+MARTIAL
+ARTS
+WEST
+BUSHRANGERS
+AUSTRALIAN
+HIGHWAYMEN
+FORMERLY
+ESCAPED
+CONVICTS
+CAPTAIN
+JAMES
+COOK
+ENGLISH
+SAILOR
+EXPLORER
+CHARLES
+BABBAGE
+BRITISH
+ANALYTICAL
+ENGINE
+FORERUNNER
+MODERN
+ROBERT
+DARWIN
+NATURALIST
+PUBLISHED
+THEORY
+EVOLUTION
+BOOK
+ENTITLED
+ORIGIN
+CRETIN
+SUFFERS
+DISEASE
+CRETINISM
+DRUIDS
+PRIESTS
+STILL
+EXISTS
+DESPITE
+EXISTENCE
+CHARLATAN
+GROUPS
+CLAIMING
+EUCLID
+ELEMENTS
+GEOMETRY
+DOWN
+TAUGHT
+NEXT
+2000
+GALILEO
+ITALIAN
+DISCOVERED
+RING
+SATURN
+JUPITER'S
+4
+MAJOR
+SATELLITES
+SUN'S
+SPOTS
+GEORGE
+ORWELL
+WRITER
+WROTE
+NINETEEN
+EIGHTY
+FARM
+HENRY
+FORD
+FOUNDER
+MOTOR
+CAR
+COMPANY
+PIONEER
+CHEAP
+WATT
+STEAM
+JOHANN
+SEBASTIAN
+BACH
+LEONARDO
+DA
+VINCI
+ARTIST
+RECORDED
+SCIENTIFIC
+STUDIES
+UNPUBLISHED
+NOTE
+BOOKS
+NEIL
+ARMSTRONG
+STEP
+ONTO
+MOON
+ORVILLE
+WRIGHT
+BROTHER
+CONTROLLED
+FLIGHT
+AEROPLANE
+PLATO
+PYTHAGORAS
+ROGUE
+IDLE
+VAGRANT
+SIR
+ISAAC
+NEWTON
+PUT
+FORWARD
+REGULATED
+SIMPLE
+MATHEMATICAL
+LAWS
+SOCRATES
+THOMAS
+ALVA
+EDISON
+CARBON
+FILAMENT
+ELECTRIC
+LAMP
+PHONOGRAPH
+WILLIAM
+SHAKESPEARE
+DRAMATIST
+ARMOUR
+PROTECTION
+WORN
+BATTLE
+MISSILE
+PROJECTED
+BOW
+BAYONET
+SHORT
+SWORD
+ATTACHED
+MUZZLE
+FIREARM
+BLOWPIPE
+TUBULAR
+WEAPON
+BLOWN
+BOKKEN
+JAPANESE
+WOODEN
+IMITATION
+SAMURAI
+KENDOKA
+WHEN
+DEMONSTRATING
+PRACTISING
+STANDARD
+KENDO
+MOVEMENTS
+BOMB
+HOLLOW
+SHELL
+FILLED
+EXPLOSIVE
+BOOMERANG
+CURVED
+ABORIGINES
+PROPELLING
+ARROWS
+BROADSWORD
+BROAD
+BLADE
+CHIEFLY
+CUTTING
+KARATE
+UNARMED
+COMBAT
+KATANA
+MERCENARY
+SOLDIER
+HIRED
+ARMY
+COUNTRY
+OFFENSIVE
+THRUSTING
+TORPEDO
+SWIMS
+TARGET
+WAKE
+TOUCH
+ERASED
+DRAW
+DIVIDING
+SCREAM
+POOR
+AFFORD
+SPEAK
+DIES
+FRUSTRATED
+SAD
+OLD
+BEFORE
+DIE
+BANK
+JUMP
+US
+TRUST
+WASN'T
+DISSAPOINTMENT
+WOULDN'T
+APPOINTMENTS
+WHAT'S
+POINT
+ARGUING
+ALONE
+MIRROR
+EYES
+CLOSED
+SHAKE
+DEVIL'S
+KIDDING
+LUCK
+CHARM
+HANGING
+OFF
+LEFT
+POLICE
+MICROSCOPE
+REVEALS
+SCOPE
+INTENTIONS
+SOMEDAY
+MONEY
+MIND
+INTERRUPT
+STRUGGLE
+COMPLETE
+THOUGHT
+RESPECT
+FORGETFUL
+YOURSELF
+PRECIOUS
+FEW
+MOMENTS
+WORST
+ENEMY
+SHARE
+OPINIONS
+WOMAN
+STRANGEST
+DREAM
+DREAMED
+KILLED
+AGAIN
+INCANDESCENT
+GAS
+GIGANTIC
+NUCLEAR
+FURNACE
--- megahal-8.6.orig/megahal.h
+++ megahal-8.6/megahal.h
@@ -140,7 +140,7 @@
 	DICTIONARY *dictionary;
 } MODEL;
 
-typedef enum { UNKNOWN, QUIT, EXIT, SAVE, DELAY, HELP, SPEECH, VOICELIST, VOICE, BRAIN } COMMAND_WORDS;
+typedef enum { UNKNOWN, QUIT, EXIT, SAVE, DELAY, HELP, SPEECH, VOICELIST, VOICE, BRAIN, QUIET} COMMAND_WORDS;
 
 typedef struct {
 	STRING word;
--- megahal-8.6.orig/Hal.pm
+++ megahal-8.6/Hal.pm
@@ -0,0 +1,101 @@
+#!/usr/bin/perl -w
+#
+# Communicate 2-way with a megahal process. This is a mess because
+# I have to handle all the IO myself using a select loop.
+#
+# Joey Hess <joey@kitenet.net> GPL 1998
+
+package Hal;
+use strict;
+use IPC::Open2;
+use Fcntl;
+use IO::Select;
+
+my $rselect=IO::Select->new();
+my $wselect=IO::Select->new();
+
+# Start megahal running and return his first line of output.
+sub start {
+	my $progname=shift || "megahal-personal";
+	
+	my $pid=open2(\*READ, \*WRITE, $progname, 
+		"--no-prompt", "--no-wrap", "--no-banner");
+	_nonblock(\*READ);
+	_nonblock(\*WRITE);
+	$rselect->add(\*READ);
+	$wselect->add(\*WRITE);
+	# Now read the first line he outputs which we will return.
+	return Hal::read();
+}
+
+# Pass a line of input to hal, returns everything he says until he expects
+# input from you again. If the input is trusted, you must pass 1 as the
+# second parameter. Otherwise, #-commands are ignored.
+sub talk { 
+	my $line=shift;
+	my $trusted=shift;
+	
+	$line=~s/^\s*//;
+	$line=~tr/ 	/ /s;
+	
+	return if (!$trusted && $line=~m/^\s*#/);
+	
+	$wselect->can_write();
+	# TODO: partial write detection &etc.
+	syswrite(WRITE,"$line\n\n",length($line)+3) || print "$!\n";
+	
+	return Hal::read();
+}
+
+# Read any pending output. There _must_ be some, or it'll block.
+sub read {
+	my $done=0;
+	my $ret='';
+	until ($done) {
+		$rselect->can_read();
+		if (sysread(READ,$_,1024)) {
+			$done=1 if (/\n/);
+			$ret.=$_;
+		}
+	}
+	return $ret;
+}
+
+# Pass a line of input to hal, in quiet mode. He listens, but does not reply.
+# See above for trusted input docs.
+sub tell {
+	my $line=shift;
+	my $trusted=shift;
+
+	$line=~s/^\s*//;
+	$line=~tr/ 	/ /s;
+
+	return if (!$trusted && $line=~m/^\s*#/);
+	
+	$wselect->can_write();
+	# TODO: partial write detection &etc.
+	syswrite(WRITE,"#quiet\n\n$line\n\n#quiet\n\n",length($line)+20) || print "$!\n";
+}
+
+# Exit cleanly.
+sub quit {
+	$wselect->can_write();
+	syswrite(WRITE,"#quit\n\n",8);
+}
+
+# Save brain.
+sub save {
+	$wselect->can_write();
+	syswrite(WRITE,"#save\n\n",8);
+}
+
+# Set a socket into nonblocking mode.
+sub _nonblock {
+	my $socket=shift;
+	my $flags= fcntl($socket, F_GETFL, 0) 
+		or die "Can't get flags for socket: $!\n";
+	fcntl($socket, F_SETFL, $flags | O_NONBLOCK)
+		or die "Can't make socket nonblocking: $!\n";
+}
+
+1
--- megahal-8.6.orig/README.debian
+++ megahal-8.6/README.debian
@@ -0,0 +1,22 @@
+Note that MegaHal depends on some data files to give it something to say. It
+can be started up without reading those files, but then it will be rather
+dumb until you teach it a lot.
+
+I've written a wrapper script for MegaHal, that creates a ~/.megahal/
+directory, and populates it with MegaHal data files. If you use this
+wrapper, each you start up MegaHal, it will resume where you left off 
+before. The wrapper is installed as /usr/bin/megahal-personal, since it lets
+you use your own personal MegaHal, which learns from you each time you use
+it.
+
+If you prefer access to the real MegaHal program, which looks for the data
+files in the current directory, run /usr/bin/megahal instead.
+
+It's also possible to set up a MegaHal that is accessed by all users of the
+system. Just make a world-writable directory and copy the MegaHal data files
+into it.
+
+Also, there is an example of an IRC bot that uses megahal in the examples/
+directory.
+
+-- Joey Hess <joeyh@debian.org>
--- megahal-8.6.orig/hal.pl
+++ megahal-8.6/hal.pl
@@ -0,0 +1,156 @@
+#!/usr/bin/perl -w
+#
+# Megahal IRC bot.
+# By Joey Hess <joey@kitenet.net> GPL 1998
+#
+# This needs the debian patched version of megahal, with #quiet command
+# and prompt to stderr output. It also needs Hal.pm and the Net::IRC module.
+#
+# Version: 1.2
+
+# Local configuration.
+my $server=shift || 'irc.debian.org';
+my $channel=shift || "#megahal";
+my $nick=shift || "megahal";
+my $ircname=shift || "Megahal";
+my $port=shift || 6667;
+my $howvocal=shift || 10; # percentage of the time it talks publically
+my $save_interval=shift || 60 * 10; # how often to save megahal's brain
+
+my @ignore=('dpkg','apt','||');	# other bots and idiots to ignore
+my $verbose=1;
+
+#
+
+use strict;
+use Net::IRC;
+use Hal;
+
+# Catch ctrl-c properly and stop megahal.
+$SIG{TERM}=sub {
+	Hal::quit();
+	exit;
+};
+$SIG{INT}=$SIG{TERM};
+
+# Set everything up.
+push @ignore,$nick;
+
+my $last_save=time();
+
+my $greet=Hal::start();
+
+my $irc=Net::IRC->new;
+my $conn=$irc->newconn(
+	Nick => $nick,
+	Server => $server,
+	Port => $port,
+	Ircname => $ircname,
+);
+$conn->add_handler('376', \&on_connect); # 376 = end of MOTD: we're connected.
+$conn->add_handler('msg', \&on_msg);
+$conn->add_handler('public', \&on_public);
+$conn->add_handler('kick', \&on_kick);
+
+$irc->start;
+
+# What to do when the bot successfully connects.
+sub on_connect {
+	my $self=shift;
+	
+	print "Joining $channel..\n" if $verbose;
+	$self->join($channel);
+	
+	# Say hi (or something vaguely along those lines).
+	print "<$nick> $greet\n";
+	$self->privmsg($channel,$greet);
+}
+
+# Listen to dialog on the channel and send it to megahal for processing.
+sub on_public {
+	my $self=shift;
+	my $event=shift;
+	
+	my $from=$event->nick;
+	
+	return if grep {$from eq $_} @ignore;
+	
+	$_=join("\n",$event->args)."\n";
+	s/[\x00-\x1f]//g; # no bold crap.
+	
+	return if /^\!.*/; # other bot talk (ie, eggdrop).
+	
+	print "<$from> $_\n" if $verbose;
+	if (/^$nick[:,;](.*)/i) {
+		if ($1) {
+			my $ret="$from: ".Hal::talk($1."\n");
+			# Respond in kind. Sometimes privatly, sometimes
+			# publically. Making it always respond publically
+			# is bad if you want to use the channel it's on
+			# for anything else, people tend to abuse the bot.
+			if (rand() * 100 <= $howvocal) {
+				print "<$nick> $ret";
+				$self->privmsg($channel,$ret);
+			}
+			else {
+				print "($nick) $ret";
+				$self->privmsg($from,$ret);
+			}
+		}
+	}
+	else {
+		# Let's try to strip out nicks and other garbage.
+		s/^.*?[:,;]//;
+		# Send it quietly - no reply is generated, to save CPU.
+		Hal::tell($_) if $_;
+	}
+
+	CheckSave();
+}
+
+
+# Listen to private messages, send to megahal for processing, and respond
+# privatly.
+sub on_msg {
+	my $self=shift;
+	my $event=shift;
+	
+	my $from=$event->nick;
+	
+	return if grep {$from eq $_} @ignore;
+	
+	$_=join("\n",$event->args)."\n";
+	s/[\x00-\x1f]*//g; # no bold crap.
+	
+	print "($from) $_\n";
+	my $ret=Hal::talk($_);
+	print "($nick) $ret";
+	$self->privmsg($from,$ret);
+
+	CheckSave();
+}
+
+# Check to see if it's been long enough since the last save for us to save
+# megahal's brain again.
+sub CheckSave {
+	if ($last_save + $save_interval < time()) {
+		$last_save=time();
+		Hal::save();		
+	}
+}
+
+# Auto-join on kick
+sub on_kick {
+	my $self=shift;
+	my $event=shift;
+	my $from=$event->nick;
+	my @stuff=$event->args;
+	my $chan=$stuff[0];
+	my $reason=$stuff[1];
+
+	$self->join($chan);
+	my $ret="$from: ".Hal::talk($reason."\n");
+	$self->privmsg($chan, $ret);
+}
+
+# All this and I still doen't understand IRC. Scaaaarey! -- Joey
--- megahal-8.6.orig/megahal-personal
+++ megahal-8.6/megahal-personal
@@ -0,0 +1,21 @@
+#!/bin/sh
+# MegaHal wrapper that makes a ~/.megahal/ directory and populates it with
+# the megahal files. This lets each user on the system have thier own
+# personal megahal.
+
+DATADIR=/usr/lib/megahal/
+
+if [ ! -d "$HOME/.megahal" ]; then
+	mkdir $HOME/.megahal
+fi
+
+cd $HOME/.megahal
+
+
+for file in aux ban grt swp trn; do
+	if [ ! -e "megahal.$file" ]; then
+		cp $DATADIR/megahal.$file .
+	fi
+done
+
+exec megahal "$@"
